### ISSUE TITLE
Warn when config contains multiple actions with the same name

### ISSUE THREAD

**Issue by Geheiner** [smart-data-lake/smart-data-lake#104]
When two actions have the same name in the config, only the last one gets taken into account and all the previous ones get discarded silently.

As most actions look quite similar, you may want to copy-paste to create them. When you do this and forget to make the action-name unique, it can be difficult to find out why some actions get ignored.

It would be nice if the smart datalake would produce a warning when two actions have the same name.

It is probably not so easy to implement this issue, since overwriting earlier occurences of config-objects by the last one is a feature of the HOCON-Language.

**Comment by swissrussdev** [comment]
I agree this would be extremely difficult to implement, as the elimination of duplicate entries is done by Typesafe. In fact, objects with duplicate keys could be distributed across multiple confs on the classpath, so a validation at file level also seems impractical.

Currently I think the best way of working around this is by generous use of reuse blocks for standard elements. That way the action blocks remain compact and it is easier to spot duplicates. Not ideal - but the best I can suggest without rewriting Typesafe!

**Comment by pgruetter** [comment]
I'm not sure if we should implement that. As noted, this is a feature of the Typesafe library:
https://github.com/lightbend/config/blob/master/HOCON.md#duplicate-keys-and-object-merging
You can explicitly merge two objects. Overwriting could also make sense in a large project where you have a hierarchy of configuration files. A warning could be unwarranted in this situation.

In any case, it's not impossible to implement. The right place is probably in the ConfigLoader, see here:
https://github.com/smart-data-lake/smart-data-lake/blob/d8f21f087e448502dfbe144a1e055e87fb48850d/src/main/scala/io/smartdatalake/config/ConfigLoader.scala#L104
At this location, you have a list of all configuration files that will be used by the ConfigParser. You could to some low level operations on all these files and check for duplicates.

**Comment by swissrussdev** [comment]
Definitely doable (default classpath loader needs some extra logic to get the files). Also needs consistent treatment of HOCON, JSON and props. 
The issue of overwriting I don't see as critical : implicit overrides of DO/Action configs are just an accident waiting to happen and are better dealt with using references to template objects. I personally see no issue with issuing a warning in this case. If someone wants to implement it, why not?

**Comment by zzeekk** [comment]
Solution implemented: 
As all files are parsed as Hocon config object before their are merged and resolved, we can check for duplicate entries for Connections, DataObjects and Actions before merging the config objects.
If duplicate entries are found, the SDLB will fail. The check can be disabled by java system property "sdl.enableCheckConfigDuplicates=false".

**Comment by Geheiner** [comment]
@zzeekk Yesterday this issue appeared again because two CopyActions had the same name within the same file due to a configuration mistake. I think that your fix only handled the case when two seperate files each contain an action with the same name.
Not sure what we can do about it though
